test/unit/backend/model_extensions/metric_test.rb

```diff
@@ -56,4 +56,32 @@ class Backend::ModelExtensions::MetricTest < ActiveSupport::TestCase
 
     metric.destroy
   end
+
+  test 'sync backend metric data only once when metric destroyed multiple times' do
+    class MetricWithFiber < ::Metric
+      def destroy_row
+        Fiber.yield
+        super
+      end
+    end
+
+    service = FactoryBot.create(:simple_service)
+    metric = MetricWithFiber.create(service: service, friendly_name: 'My metric', unit: 'hits')
+    metric_id = metric.id
+
+    ::BackendMetricWorker.expects(:sync).with(service.backend_id, metric_id, metric.system_name).once
```

Comment:
This should be red to have it reproduced and it is green 😞 Imo we should add 'loggers' instead of trying to fix it and then when it happens again (which happens often), check the logs and see what happened 😄

Reply:
I figured it out. It is indeed a race condition problem:
- 2 threads are trying to destroy a metric – T1 and T2
- T1 succeeds but it's interrupted right after `ActiveRecord::Persistence#destroy` sets `@destroyed = true` and yet before T2 reaches inside the same method the check `if persisted?`
- T2 gets `persisted? => false` inside `ActiveRecord::Persistence#destroy` due to https://apidock.com/rails/v4.2.7/ActiveRecord/Core/sync_with_transaction_state, thus making `@_trigger_destroy_callback = true`
- Since `m.destroyed? == true` in https://github.com/3scale/porta/blob/8e55bade8619287b5a9b9ae7defc0d3a49cd3a0d/app/lib/backend/model_extensions/metric.rb#L9, both T1 and T2 invoke the callback